File: src/ObjectManager/Object.Tes/FilePacks/EsmFile+Extensions.cs
using OA.Core;
using OA.Tes.FilePacks.Records;
using System;
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using System.Text;

namespace OA.Tes.FilePacks
{
    partial class EsmFile
    {
        // TES3
        internal Dictionary<string, Record> _MANYsById;
        Dictionary<long, LTEXRecord> _LTEXsById;
        Dictionary<Vector3Int, LANDRecord> _LANDsById;
        Dictionary<Vector3Int, CELLRecord> _CELLsById;
        Dictionary<string, CELLRecord> _CELLsByName;

        // TES4
        Dictionary<uint, Tuple<WRLDRecord, RecordGroup[]>> _WRLDsById;
        Dictionary<string, LTEXRecord> _LTEXsByEid;

        void Process()
        {
            if (Format == GameFormatId.TES3)
            {
                var manyGroups = new List<Record>[] { Groups.ContainsKey("STAT") ? Groups["STAT"].Load() : null };
                _MANYsById = manyGroups.SelectMany(x => x).Cast<IHaveEDID>().Where(x => x != null).ToDictionary(x => x.EDID.Value, x => (Record)x);
                _LTEXsById = Groups["LTEX"].Load().Cast<LTEXRecord>().ToDictionary(x => x.INTV.Value);
                _LANDsById = Groups["LAND"].Load().Cast<LANDRecord>().ToDictionary(x => x.GridId);
                var cells = Groups["CELL"].Load().Cast<CELLRecord>().ToList();
                foreach (var cell in cells)
                    cell.GridId = new Vector3Int(cell.XCLC.Value.GridX, cell.XCLC.Value.GridY, !cell.IsInterior ? 0 : -1);
                _CELLsById = cells.Where(x => !x.IsInterior).ToDictionary(x => x.GridId);
                _CELLsByName = cells.Where(x => x.IsInterior).ToDictionary(x => x.EDID.Value);
                return;
            }
            var wrldsByLabel = Groups["WRLD"].GroupsByLabel;
            _WRLDsById = Groups["WRLD"].Load().Cast<WRLDRecord>().ToDictionary(x => x.Id, x =>
            {
                wrldsByLabel.TryGetValue(BitConverter.GetBytes(x.Id), out var wrlds);
                return new Tuple<WRLDRecord, RecordGroup[]>(x, wrlds);
            });
            _LTEXsByEid = Groups["LTEX"].Load().Cast<LTEXRecord>().ToDictionary(x => x.EDID.Value);
        }

        public LTEXRecord FindLTEXRecord(int index)
        {
            if (Format == GameFormatId.TES3)
            {
                _LTEXsById.TryGetValue(index, out var ltex);
                return ltex;
            }
            throw new NotImplementedException();
        }

        public LANDRecord FindLANDRecord(Vector3Int cellId)
        {
            if (Format == GameFormatId.TES3)
            {
                _LANDsById.TryGetValue(cellId, out var land);
                return land;
            }
            var world = _WRLDsById[(uint)cellId.z];
            foreach (var wrld in world.Item2)
                foreach (var cellBlock in wrld.EnsureWrldAndCell(cellId))
                    if (cellBlock.LANDsById.TryGetValue(cellId, out var land))
                        return land;
            return null;
        }

        public CELLRecord FindCellRecord(Vector3Int cellId)
        {
            if (Format == GameFormatId.TES3)
            {
                _CELLsById.TryGetValue(cellId, out var cell);
                return cell;
            }
            var world = _WRLDsById[(uint)cellId.z];
            foreach (var wrld in world.Item2)
                foreach (var cellBlock in wrld.EnsureWrldAndCell(cellId))
                    if (cellBlock.CELLsById.TryGetValue(cellId, out var cell))
                        return cell;
            return null;
        }

        public CELLRecord FindCellRecordByName(int world, int cellId, string cellName)
        {
            if (Format == GameFormatId.TES3)
            {
                _CELLsByName.TryGetValue(cellName, out var cell);
                return cell;
            }
            throw new NotImplementedException();
        }
    }

    partial class RecordGroup
    {
        internal HashSet<byte[]> _ensureCELLsByLabel;
        internal Dictionary<Vector3Int, CELLRecord> CELLsById;
        internal Dictionary<Vector3Int, LANDRecord> LANDsById;

        public RecordGroup[] EnsureWrldAndCell(Vector3Int cellId)
        {
            var cellBlockX = (short)(cellId.x >> 5);
            var cellBlockY = (short)(cellId.y >> 5);
            var cellBlockId = new byte[4];
            Buffer.BlockCopy(BitConverter.GetBytes(cellBlockY), 0, cellBlockId, 0, 2);
            Buffer.BlockCopy(BitConverter.GetBytes(cellBlockX), 0, cellBlockId, 2, 2);
            Load();
            if (GroupsByLabel.TryGetValue(cellBlockId, out var cellBlocks))
                return cellBlocks.Select(x => x.EnsureCell(cellId)).ToArray();
            return null;
        }

        //  = nxn[nbits] + 4x4[2bits] + 8x8[3bit]
        public RecordGroup EnsureCell(Vector3Int cellId)
        {
            if (_ensureCELLsByLabel == null)
                _ensureCELLsByLabel = new HashSet<byte[]>(ByteArrayComparer.Default);
            var cellBlockX = (short)(cellId.x >> 5);
            var cellBlockY = (short)(cellId.y >> 5);
            var cellSubBlockX = (short)(cellId.x >> 3);
            var cellSubBlockY = (short)(cellId.y >> 3);
            var cellSubBlockId = new byte[4];
            Buffer.BlockCopy(BitConverter.GetBytes(cellSubBlockY), 0, cellSubBlockId, 0, 2);
            Buffer.BlockCopy(BitConverter.GetBytes(cellSubBlockX), 0, cellSubBlockId, 2, 2);
            if (_ensureCELLsByLabel.Contains(cellSubBlockId))
                return this;
            Load();
            if (CELLsById == null)
                CELLsById = new Dictionary<Vector3Int, CELLRecord>();
            if (LANDsById == null && cellId.z >= 0)
                LANDsById = new Dictionary<Vector3Int, LANDRecord>();
            if (GroupsByLabel.TryGetValue(cellSubBlockId, out var cellSubBlocks))
            {
                // find cell
                var cellSubBlock = cellSubBlocks.Single();
                cellSubBlock.Load(true);
                foreach (var cell in cellSubBlock.Records.Cast<CELLRecord>())
                {
                    cell.GridId = new Vector3Int(cell.XCLC.Value.GridX, cell.XCLC.Value.GridY, !cell.IsInterior ? cellId.z : -1);
                    CELLsById.Add(cell.GridId, cell);
                    // find children
                    if (cellSubBlock.GroupsByLabel.TryGetValue(BitConverter.GetBytes(cell.Id), out var cellChildren))
                    {
                        var cellChild = cellChildren.Single();
                        var cellTemporaryChildren = cellChild.Groups.Single(x => x.Headers.First().GroupType == Header.HeaderGroupType.CellTemporaryChildren);
                        foreach (var land in cellTemporaryChildren.Records.Cast<LANDRecord>())
                        {
                            land.GridId = new Vector3Int(cell.XCLC.Value.GridX, cell.XCLC.Value.GridY, !cell.IsInterior ? cellId.z : -1);
                            LANDsById.Add(land.GridId, land);
                        }
                    }
                }
                _ensureCELLsByLabel.Add(cellSubBlockId);
                return this;
            }
            return null;
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/EsmFile+Extensions.swift
//
//  EsmFile+Extensions.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import Foundation
import simd

extension EsmFile {
    func process() {
        guard format != .TES3 else {
            let manyGroups = [groups!["STAT"]?.load(), nil]
            _MANYsById = Dictionary(uniqueKeysWithValues: manyGroups.compactMap { $0 }.flatMap { $0 }.compactMap { $0 as? IHaveEDID }
                .map { (key: $0.EDID, value: $0 as! Record) })
            _LTEXsById = Dictionary(uniqueKeysWithValues: groups!["LTEX"]!.load().map { $0 as! LTEXRecord }
                .map { (key: $0.INTV, value: $0) })
            let lands = groups!["LAND"]!.load().map { $0 as! LANDRecord }
            for land in lands {
                land.gridId = int3(Int32(land.INTV.cellX), Int32(land.INTV.cellY), 0)
            }
            _LANDsById = Dictionary(uniqueKeysWithValues: lands
                .map { (key: $0.gridId, value: $0) })
            let cells = groups!["CELL"]!.load().map { $0 as! CELLRecord }
            for cell in cells {
                cell.gridId = int3(Int32(cell.XCLC!.gridX), Int32(cell.XCLC!.gridY), !cell.isInterior ? 0 : -1)
            }
            _CELLsById = Dictionary(uniqueKeysWithValues: cells.filter { !$0.isInterior }
                .map { (key: $0.gridId, value: $0) })
            _CELLsByName = Dictionary(uniqueKeysWithValues: cells.filter { $0.isInterior }
                .map { (key: $0.EDID, value: $0) })
            // temp
            for a in _CELLsById.keys {
                let b = _LANDsById[a]
                if b?.VTEX != nil {
                    debugPrint("\(a)")
                }
            }
            return
        }
        let wrldsByLabel = groups!["WRLD"]!.groupsByLabel!
        _WRLDsById = Dictionary(uniqueKeysWithValues: groups!["WRLD"]!.load().map { $0 as! WRLDRecord }
            .map { (key: $0.id, value: (x: $0, y: wrldsByLabel[$0.id]!)) })
        _LTEXsByEid = Dictionary(uniqueKeysWithValues: groups!["LTEX"]!.load().map { $0 as! LTEXRecord }
            .map { (key: $0.EDID, value: $0) })
    }

    func findLTEXRecord(index: Int64) -> LTEXRecord? {
        guard format != .TES3 else {
            return _LTEXsById[index]
        }
        fatalError("NotImplemented")
    }

    func findLANDRecord(cellId: int3) -> LANDRecord? {
        guard format != .TES3 else {
            return _LANDsById[cellId]
        }
        let world = _WRLDsById[UInt32(cellId.z)]!
        for wrld in world.y {
            for cellBlock in wrld.ensureWrldAndCell(cellId: cellId)! {
                if let land = cellBlock.LANDsById[cellId] {
                     return land
                }
            }
        }
        return nil
    }

    public func findCellRecord(cellId: int3) -> CELLRecord? {
        guard format != .TES3 else {
            return _CELLsById[cellId]
        }
        let world = _WRLDsById[UInt32(cellId.z)]!
        for wrld in world.y {
            for cellBlock in wrld.ensureWrldAndCell(cellId: cellId)! {
                if let cell = cellBlock.CELLsById[cellId] {
                    return cell
                }
            }
        }
        return nil
    }
    
    public func findCellRecordByName(world: Int32, cellId: Int, cellName: String) -> CELLRecord? {
        guard format != .TES3 else {
            return _CELLsByName[cellName]
        }
        fatalError("NotImplemented")
    }
}

extension RecordGroup {    
    public func ensureWrldAndCell(cellId: int3) -> [RecordGroup]? {
        let cellBlockX = Int16(cellId.x >> 5)
        let cellBlockY = Int16(cellId.y >> 5)
        let cellBlockId = UInt32((cellBlockY << 8) | cellBlockX)
        load()
        guard let cellBlocks = groupsByLabel![cellBlockId] else {
            return nil
        }
        return cellBlocks.compactMap { $0.ensureCell(cellId) }
    }
    
    func ensureCell(_ cellId: int3) -> RecordGroup? {
        if _ensureCELLsByLabel == nil {
            _ensureCELLsByLabel = Set()
        }
//        let cellBlockX = Int16(cellId.x >> 5)
//        let cellBlockY = Int16(cellId.y >> 5)
        let cellSubBlockX = Int16(cellId.x >> 3)
        let cellSubBlockY = Int16(cellId.y >> 3)
        let cellSubBlockId = UInt32((cellSubBlockY << 8) | cellSubBlockX)
        guard !_ensureCELLsByLabel!.contains(cellSubBlockId) else {
            return self
        }
        load()
        if CELLsById == nil {
            CELLsById = [int3 : CELLRecord]()
        }
        if LANDsById == nil && cellId.z >= 0 {
            LANDsById = [int3 : LANDRecord]()
        }
        guard let cellSubBlocks = groupsByLabel![cellSubBlockId] else {
            return nil
        }
        // find cell
        let cellSubBlock = cellSubBlocks.first!
        cellSubBlock.load(loadAll: true)
        for cell in (cellSubBlock.records.map { $0 as! CELLRecord }) {
            cell.gridId = int3(Int32(cell.XCLC!.gridX), Int32(cell.XCLC!.gridY), !cell.isInterior ? cellId.z : -1)
            CELLsById[cell.gridId] = cell
            // find children
            if let cellChildren = cellSubBlock.groupsByLabel![cell.id] {
                let cellChild = cellChildren.first!
                let cellTemporaryChildren = (cellChild.groups!.filter { $0.headers.first!.groupType == .cellTemporaryChildren }).first!
                for land in (cellTemporaryChildren.records.map { $0 as! LANDRecord }) {
                    land.gridId = int3(cell.XCLC!.gridX, cell.XCLC!.gridY, !cell.isInterior ? cellId.z : -1);
                    LANDsById[land.gridId] = land
                }
            }
        }
        _ensureCELLsByLabel.insert(cellSubBlockId)
        return self
    }
}
